token-metadata/program/src/state.rs

```diff
@@ -776,3 +802,278 @@ impl EditionMarker {
         Ok(())
     }
 }
+
+#[cfg(test)]
+
+mod metadata_deserialization {
+    use solana_sdk::{signature::Keypair, signer::Signer};
+
+    use super::*;
+    use crate::{
+        deser::tests::expected_pesky_metadata,
+        state::{Edition, MasterEditionV2},
+    };
+    pub use crate::{state::Creator, utils::puff_out_data_fields};
+    use std::io::ErrorKind;
+
+    fn pad_metadata_length(metadata: &mut Vec<u8>) {
+        let padding_length = MAX_METADATA_LEN - metadata.len();
+        metadata.extend(vec![0; padding_length]);
+    }
+
+    #[test]
+    fn successfully_deserialize_metadata() {
+        // Setup for BorshDeserialze `deserialize` test.
+        let expected_metadata = expected_pesky_metadata();
+
+        let mut buf = Vec::new();
+        expected_metadata.serialize(&mut buf).unwrap();
+        pad_metadata_length(&mut buf);
+
+        // Setup for `from_account_info` test.
+        let pubkey = Keypair::new().pubkey();
+        let owner = &ID;
+        let mut lamports = 1_000_000_000;
+        let mut data = buf.clone();
+
+        let md_account_info = AccountInfo::new(
+            &pubkey,
+            false,
+            true,
+            &mut lamports,
+            &mut data,
+            &owner,
+            false,
+            1_000_000_000,
+        );
+
+        // Both deserialization methods should succeed.
+        let metadata = Metadata::deserialize(&mut buf.as_ref()).unwrap();
+        assert_eq!(metadata.key, Key::MetadataV1);
+        assert_eq!(metadata, expected_metadata);
+
+        let md = Metadata::from_account_info(&md_account_info).unwrap();
+        assert_eq!(md.key, Key::MetadataV1);
+        assert_eq!(md, expected_metadata);
+    }
+
+    #[test]
+    fn fail_to_deserialize_metadata_with_wrong_owner() {
+        let expected_metadata = expected_pesky_metadata();
+
+        let mut buf = Vec::new();
+        expected_metadata.serialize(&mut buf).unwrap();
+        pad_metadata_length(&mut buf);
+
+        let pubkey = Keypair::new().pubkey();
+        let invalid_owner = Keypair::new().pubkey();
+        let mut lamports = 1_000_000_000;
+        let mut data = buf.clone();
+
+        let md_account_info = AccountInfo::new(
+            &pubkey,
+            false,
+            true,
+            &mut lamports,
+            &mut data,
+            &invalid_owner,
+            false,
+            1_000_000_000,
+        );
+
+        // `from_account_info` should not succeed because this account is not owned
+        // by `token-metadata` program.
+        let error = Metadata::from_account_info(&md_account_info).unwrap_err();
+        assert_eq!(error, MetadataError::IncorrectOwner.into());
+
+        // `deserialize` should succeed because it just deserializes bytes, with no account context.
+        let metadata = Metadata::deserialize(&mut buf.as_ref()).unwrap();
+        assert_eq!(metadata, expected_metadata);
+    }
+
+    #[test]
+    fn fail_to_deserialize_master_edition() {
```

Comment:
Not super familiar with all the account types, but what about other ones from the `Key` enum such as `MasterEditionV1` and `ReservationListV2`?

Or are the tests sufficient with `MasterEditionV2`, `EditionV1`, `UseAuthorityRecord`, and `CollectionAuthorityRecord`, `EditionMarker`.

Just curious why that subset?  Not sure if the tests are meant to be exhaustive and if they are or not.

Reply:
`ReserverationV1`, `ReservationV2`, and `MasterEditionV1` are not widely in use (or at all?) as they're from old really early versions of the code. I figured the subset of the most common was sufficient to demonstrate the code works as expected but it wouldn't be hard to make the list exhaustive if you think I should.

Reply:
I think its fine, proving a couple cases showed you got the business logic of the check correct.  Any more is probably overkill, like proving that enums work properly.